text/0517-io-os-reform.md

```diff
@@ -473,12 +635,289 @@ throughout IO, we can go on to explore the modules in detail.
+in an extension trait with a blanket `impl`.
+
+Note that the same `ByRef` type is used for both `Read` and `Write`
+-- and this RFC proposes to use it for `std::iter` as well. The
+insight is that there is no difference between the *type* used for
+by-ref adapters in any of these cases; what changes is just which
+trait defers through it. So, we propose to add the following to `core::borrow`:
+
+```rust
+pub struct ByRef<'a, Sized? T:'a> {
+    pub inner: &'a mut T
+}
+```
+
+which will allow `impl`s like the following in `core::io`:
+
+```rust
+impl<'a, W: Write> Write for ByRef<'a, W> {
+    #[inline]
+    fn write(&mut self, buf: &[u8]) -> Result<uint, W::Err> { self.inner.write(buf) }
+
+    #[inline]
+    fn flush(&mut self) -> Result<(), W::Err> { self.inner.flush() }
+}
+```
+
+#### Free functions
+[Free functions]: #free-functions
+
+The current `std::io::util` module also includes a number of primitive
+readers and writers, as well as `copy`. These are updated as follows:
+
+```rust
+// A reader that yields no bytes
+fn empty() -> Empty; // in theory just returns `impl Reader`
+
+// A reader that yields `byte` repeatedly (generalizes today's ZeroReader)
+fn repeat(byte: u8) -> Repeat;
+
+// A writer that ignores the bytes written to it (/dev/null)
+fn sink() -> Sink;
+
+// Copies all data from a Reader to a Writer
+pub fn copy<E, R, W>(r: &mut R, w: &mut W) -> Result<(), E> where
+    R: Read<Err = E>,
+    W: Write<Err = E>
+```
+
+#### Seeking
+[Seeking]: #seeking
+
+The seeking infrastructure is largely the same as today's, except that
+`tell` is renamed to follow the RFC's design principles and the `seek`
+signature is refactored with more precise types:
+
+```rust
+pub trait Seek {
+    type Err;
+    fn position(&self) -> Result<u64, Err>;
+
+    // returns the new position after seeking
+    fn seek(&mut self, pos: SeekPos) -> Result<u64, Err>;
+}
+
+pub enum SeekPos {
+    FromStart(u64),
+    FromEnd(i64),
+    FromCur(i64),
+}
+```
+
+#### Buffering
+[Buffering]: #buffering
+
+The current `Buffer` trait will be renamed to `BufferedRead` for
+clarity (and to open the door to `BufferedWrite` at some later
+point):
+
+```rust
+pub trait BufferedRead: Read {
+    fn fill_buf(&mut self) -> Result<&[u8], Self::Err>;
+    fn consume(&mut self, amt: uint);
+
+    fn read_until(&mut self, byte: u8) -> ReadUntil { ... }
+}
+
+pub trait BufferedReadExt: BufferedRead {
+    fn lines(&mut self) -> Lines<Self, Self::Err> { ... };
+}
+```
+
+where `ReadUntil: Iterator<Result<Vec<u8>, Self::Err>>`.  In addition,
+`read_line` is removed in favor of the `lines` iterator, and
+`read_char` is removed in favor of the `chars` iterator (now on
+`ReadExt`). As with all nonatomic convenience methods, these methods
+do not offer a means of recovering from a transient errors; one should
+use the lower-level methods in such cases.
+
+The `BufferedReader`, `BufferedWriter` and `BufferedStream` types stay
```

Comment:
I feel like it's somewhat unfortunate that the name of these types is so long. Typing `BufferedReader::new(try!(File::open("foo")))` gets somewhat weary after awhile. Down below there's a "new trend" with `Cursor<T>` where the same structure is used for reading/writing/unifying these concepts. I'd love to have `Buffer<T>` where `T` is either a `Read` or `Write` instance, but I'm not totally sure it would work out. With only one buffer you'd have to always `read` or always `write` the type, but there wouldn't be a guarantee either way.

In the end, `BufferedReader` isn't necessarily the worst thing in the world though!